### PR TITLE
Add parser support for DECLARE, FETCH and CLOSE

### PR DESCRIPTION
**Pull request by mfussenegger**
## Summary of the changes / Why this improves CrateDB

Subset of https://github.com/crate/crate/pull/12597
Made some small modifications:

- Direction of fetch is normalized into ABSOLUTE or RELATIVE
- Shorter class names
- DECLARE/FETCH form no hierarchy

(Not sure if this should go into a feature branch. Imho there's no harm merging this directly to master)

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Review by BaurzhanSakhariev — APPROVED** [review]

File: libs/sql-parser/src/main/antlr/SqlBase.g4 [hunk 1]
```diff
@@ -746,7 +773,8 @@ nonReserved
     | REPLACE | RETURNING | SWAP | GC | DANGLING | ARTIFACTS | DECOMMISSION | LEADING | TRAILING | BOTH | TRIM
     | CURRENT_SCHEMA | PROMOTE | CHARACTER | VARYING | SUBSTRING
     | DISCARD | PLANS | SEQUENCES | TEMPORARY | TEMP | METADATA
-    | PUBLICATION | SUBSCRIPTION | ENABLE | DISABLE | CONNECTION
+    | PUBLICATION | SUBSCRIPTION | ENABLE | DISABLE | CONNECTION | DECLARE | CURSOR | HOLD | FORWARD | BACKWARD
```
Comment (matriv) [inline libs/sql-parser/src/main/antlr/SqlBase.g4:776]: not important, and not for this PR, do you think it would be nice to sort these lists alphabetically at some point?

File: libs/sql-parser/src/main/java/io/crate/sql/SqlFormatter.java [hunk 2]
```diff
@@ -1219,6 +1223,64 @@ public Void visitSetSessionAuthorizationStatement(SetSessionAuthorizationStateme
...
+                    builder.append(fetch.count());
+                    builder.append(" ");
+                }
+            } else {
+                if (count >= 0) {
+                    builder.append("FORWARD ");
+                } else {
+                    builder.append("BACKWARD ");
+                }
+                if (count == Long.MAX_VALUE || count == Long.MAX_VALUE * -1) {
```
Comment (matriv) [inline libs/sql-parser/src/main/java/io/crate/sql/SqlFormatter.java:1264]: why not use `Long.MIN_VALUE` here?
```
System.out.println(Long.MIN_VALUE);
System.out.println(Long.MAX_VALUE * -1);
=>
-9223372036854775808
-9223372036854775807
```
Is it specified by the CURSOR requirements?

File: libs/sql-parser/src/main/java/io/crate/sql/tree/Close.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    @Override
+    public <R, C> R accept(AstVisitor<R, C> visitor, C context) {
+        return visitor.visitClose(this, context);
+    }
+
+    @Override
+    public int hashCode() {
+        return cursorName == null ? 0 : cursorName.hashCode();
```
Comment (matriv) [inline libs/sql-parser/src/main/java/io/crate/sql/tree/Close.java:51]: Why not use `Objects.hashCode(cursorName)` ?

File: libs/sql-parser/src/main/java/io/crate/sql/tree/Fetch.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        if (getClass() != obj.getClass()) {
+            return false;
+        }
+        Fetch other = (Fetch) obj;
+        return count == other.count && Objects.equals(cursorName, other.cursorName) && scrollMode == other.scrollMode;
+    }
+
+    @Override
+    public String toString() {
+        return "Fetch{count=" + count + ", cursorName=" + cursorName + ", moveMode=" + scrollMode + "}";
```
Comment (matriv) [inline libs/sql-parser/src/main/java/io/crate/sql/tree/Fetch.java:107]: why `moveMode` and not `scrollMode`?

File: server/src/test/java/io/crate/analyze/SelectStatementAnalyzerTest.java [hunk 5]
```diff
@@ -3042,4 +3042,25 @@ public void testSubscriptExpressionFromUnionAll() throws Exception {
         assertThat(analyzed.outputs().size(), is(1));
         assertThat(analyzed.outputs().get(0), isFunction("subscript", isField("obj"), isLiteral("unknown")));
     }
+
+    @Test
+    public void test_declare_is_not_supported() {
+        var executor = SQLExecutor.builder(clusterService).build();
```
Comment (matriv) [inline server/src/test/java/io/crate/analyze/SelectStatementAnalyzerTest.java:3048]: Also not part of the PR, could we extract the executor and make ie available for all (or at least most) tests?

File: libs/sql-parser/src/main/java/io/crate/sql/SqlFormatter.java (hunk 2 above)
Comment (mfussenegger) [reply libs/sql-parser/src/main/java/io/crate/sql/SqlFormatter.java:1264]: Makes it easier to use the same value as sentinel for ALL

File: libs/sql-parser/src/main/java/io/crate/sql/tree/Close.java (hunk 3 above)
Comment (mfussenegger) [reply libs/sql-parser/src/main/java/io/crate/sql/tree/Close.java:51]: Causes array allocation

File: server/src/test/java/io/crate/analyze/SelectStatementAnalyzerTest.java (hunk 5 above)
Comment (mfussenegger) [reply server/src/test/java/io/crate/analyze/SelectStatementAnalyzerTest.java:3048]: Can do it in the follow up PR that adds the implementation. 

In general I kinda prefer the pattern of making it local to the test as it gives you more flexibility defining the tables without adding overhead due multiple table definitions. But I guess in this case it's fine to have a single setup

File: libs/sql-parser/src/main/java/io/crate/sql/tree/Fetch.java (hunk 4 above)
Comment (mfussenegger) [reply libs/sql-parser/src/main/java/io/crate/sql/tree/Fetch.java:107]: Forgot to update after a rename

File: libs/sql-parser/src/main/java/io/crate/sql/tree/Close.java (hunk 3 above)
Comment (matriv) [reply libs/sql-parser/src/main/java/io/crate/sql/tree/Close.java:51]: not `Objects.hash(Object... values)` but `Objects.hashCode()`, but anyway not important at all.

File: server/src/test/java/io/crate/analyze/SelectStatementAnalyzerTest.java (hunk 5 above)
Comment (matriv) [reply server/src/test/java/io/crate/analyze/SelectStatementAnalyzerTest.java:3048]: Not insisting, I was mostly asking if there is a reasoning/benefit of this pattern.

**Review by matriv — APPROVED** [review]

File: libs/sql-parser/src/main/java/io/crate/sql/tree/Fetch.java (hunk 4 above)
Comment (matriv) [reply libs/sql-parser/src/main/java/io/crate/sql/tree/Fetch.java:107]: reminder: update the name before merging, thx!

File: libs/sql-parser/src/main/java/io/crate/sql/tree/Close.java (hunk 3 above)
Comment (mfussenegger) [reply libs/sql-parser/src/main/java/io/crate/sql/tree/Close.java:51]: Oh right, changing it

**Comment by mfussenegger** [comment]
retest this please

`testTransientErrorsDuringRecoveryAreRetried` failed

```
java.lang.AssertionError: timed out waiting for green state
	at __randomizedtesting.SeedInfo.seed([23F39644E5EB506D:4A197A1B23EDB8F9]:0)
	at org.junit.Assert.fail(Assert.java:89)
	at org.elasticsearch.test.IntegTestCase.ensureColor(IntegTestCase.java:782)
	at org.elasticsearch.test.IntegTestCase.ensureGreen(IntegTestCase.java:736)
	at org.elasticsearch.test.IntegTestCase.ensureGreen(IntegTestCase.java:725)
	at org.elasticsearch.indices.recovery.IndexRecoveryIT.testTransientErrorsDuringRecoveryAreRetried(IndexRecoveryIT.java:818)
```
